inductiva/_cli/ansi_pager.py

```diff
@@ -0,0 +1,464 @@
+            self._buffer.clear()
+            self._clear_screen()
+            self._render_footer()
+
+    def run(self) -> None:
+        """Run the navigation loop"""
+
+        stdin = sys.stdin.fileno()
+        tattr = termios.tcgetattr(stdin)
+        tty.setcbreak(stdin, termios.TCSANOW)
+
+        scroll = {
+            "\033[A": partial(self._scroll, self.SCROLL_UP, False),
+            "\033[B": partial(self._scroll, self.SCROLL_DOWN, False),
+            "\033[C": partial(self._scroll, self.SCROLL_DOWN, True),
+            "\033[D": partial(self._scroll, self.SCROLL_UP, True),
+            " ": partial(self._scroll, self.SCROLL_DOWN, True),
+            "b": partial(self._scroll, self.SCROLL_UP, True),
+        }
+
+        try:
+            buf = code = ""
+            while True:
+                buf = sys.stdin.read(1)
+                if buf == "q":
+                    break
+                if action := scroll.get(buf):
+                    action()
+                elif buf == "\033":
+                    code = buf
+                elif buf in "[ABCD":
+                    code += buf
+                else:
+                    code = ""
+                    continue
+
+                if action := scroll.get(code):
+                    action()
+                    code = ""
+
+        except KeyboardInterrupt:
+            termios.tcsetattr(stdin, termios.TCSANOW, tattr)
+
+    def write(self, text: str) -> int:
+        """Write string to the file-like object.
+
+        Returns the total number of lines in the display.
+        Note: The display is only updated after a call to the `flush` method.
+        """
+        with self._lock:
+            start = len(self._buffer)
+            lines = text.splitlines()
+            self._buffer.extend(lines)
+            self._render_body(start)
+            self._render_footer()
+            return len(self._buffer) - start
+
+    def writelines(self, lines: Iterable[AnyStr]) -> None:
+        """Write a list of lines to the file-like object.
+
+        Line separators are not added, so it is usual for each of the
+        lines provided to have a line separator at the end.
+        Note: The display is only updated after a call to the `flush` method.
+        """
+        with self._lock:
+            start = len(self._buffer)
+            self._buffer.extend(x for line in lines for x in line.splitlines())
+            self._render_body(start)
+            self._render_footer()
+
+    def flush(self) -> None:
+        """Flush the content of the file-like object to the display."""
+        sys.stdout.flush()
+
+    def close(self) -> None:
+        """Close the file-like object.
+
+        The method closes the file-like object and resets the terminal
+        to its previous state."""
+        self._resizer.stop()
+        self._resizer.join()
+        self._exit_alt_screen()
+        os.system("stty sane")  # Required to return terminal to normal.
```

Comment:
Which terminal are you using? This error appears to me in bash, I have tried in zsh and I don't have the problem. It seems that the terminal needs to be reset. For example, after commenting this and executing one prompt that creates a pager, I get the error. If I type reset (even though invisibly) it goes back to how it was.

In fact, if it `os.system("reset")` it refactors back again, but it takes longer and clears the terminal fully.

Reply:
I've tried it with zsh and bash in iterm2, terminal.app and vscode's terminal.

Reply:
To be fair, I don't understand the reason to not work on my computer...
But, to be completely sure, I will comment this on QA so that we can verify if it works for other people.